uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/POMServiceImpl.java

```diff
@@ -153,74 +163,118 @@ public Path save(final Path path,
-                    }
-                }
+        private void savePOM(final Path path,
+                             final POM content,
+                             final Metadata metadata) throws IOException, XmlPullParserException {
+            final Optional<Module> oldModule = getModuleIfNameChanged(path,
+                                                                      content);
+
+            if (metadata == null) {
+                ioService.write(Paths.convert(path),
+                                pomContentHandler.toString(content,
+                                                           loadPomXMLString(path)));
+            } else {
+                ioService.write(Paths.convert(path),
+                                pomContentHandler.toString(content,
+                                                           loadPomXMLString(path)),
+                                metadataService.setUpAttributes(path,
+                                                                metadata));
+            }
+
+            if (oldModule.isPresent()) {
+                updates.add(() -> renameModuleEvent.fire(new RenameModuleEvent(oldModule.get(),
+                                                                               moduleService.resolveModule(path))));
             }
         }
-    }
 
-    private Path save(final Path path,
-                      final POM content,
-                      final Metadata metadata) throws IOException, XmlPullParserException {
-        if (metadata == null) {
-            save(path,
-                 content);
-        } else {
-            ioService.write(Paths.convert(path),
-                            pomContentHandler.toString(content,
-                                                       loadPomXMLString(path)),
-                            metadataService.setUpAttributes(path,
-                                                            metadata));
+        private Optional<Module> getModuleIfNameChanged(final Path path,
+                                                        final POM content) {
+            if (!load(path).getName().equals(content.getName())) {
+                return Optional.of(moduleService.resolveModule(path));
+            } else {
+                return Optional.empty();
+            }
         }
 
-        return path;
-    }
+        private void saveSubModulePOMs() throws IOException, XmlPullParserException {
+
+            if (content.isMultiModule() &&
+                    content.getModules() != null) {
+
+                for (final String module : content.getModules()) {
+
+                    final org.uberfire.java.nio.file.Path childPath = Paths.convert(path).getParent().resolve(module).resolve("pom.xml");
+
+                    if (ioService.exists(childPath)) {
+                        final POM child = load(Paths.convert(childPath));
+                        if (child != null) {
+                            child.setParent(content.getGav());
```

Comment:
Id save ` GAV parentGav = content.getGav();` for clarity and not to repeat the call 3 times. Also looking at how deep this is nested, maybe extracting this out to helper would help readability too

```
content.getModules()
    .forEach(childModule -> saveGavChange(parentGav, childModule));

```